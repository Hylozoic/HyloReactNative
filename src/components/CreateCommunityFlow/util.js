export const slugValidatorRegex = /^[0-9a-z-]{2,40}$/
export const invalidSlugMessage = 'URLs must have between 2 and 40 characters, and can only have lower case letters, numbers, and dashes.'
