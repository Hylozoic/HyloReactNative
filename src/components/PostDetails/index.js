import component from './PostDetails'
import connector from './PostDetails.connector'
export default connector(component)
