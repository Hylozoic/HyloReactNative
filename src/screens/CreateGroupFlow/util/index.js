export const slugValidatorRegex = /^[0-9a-z-]{2,40}$/
export const invalidSlugMessage = 'URLs must have between 2 and 40 characters, and can only have lower case letters, numbers, and dashes.'

export function formatDomainWithUrl (groupDomain) {
  if (!groupDomain) return
  let formattedDomain = groupDomain.replace(/hylo\.com\/groups\/?/, '')
  if (formattedDomain !== '') {
    formattedDomain = 'hylo.com/groups/' + formattedDomain
  }
  return formattedDomain
}

export function removeDomainFromURL (groupDomain) {
  if (!groupDomain) return
  return groupDomain.replace('hylo.com/groups/', '')
}
