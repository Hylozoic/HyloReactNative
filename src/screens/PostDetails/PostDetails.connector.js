import { connect } from 'react-redux'
import { get } from 'lodash/fp'
import { isModalScreen, modalScreenName } from 'navigation/linking/helpers'
import fetchPost from 'store/actions/fetchPost'
import getCurrentGroup from 'store/selectors/getCurrentGroup'
import { getPresentedPost } from 'store/selectors/getPost'
import getRouteParam from 'store/selectors/getRouteParam'
import getMe from 'store/selectors/getMe'
import joinProject from 'store/actions/joinProject'
import leaveProject from 'store/actions/leaveProject'
import respondToEvent from 'store/actions/respondToEvent'

export function mapStateToProps (state, props) {
  const currentUser = getMe(state, props)
  const currentGroup = getCurrentGroup(state, props)
  const postId = getRouteParam('id', props.route)
  const post = getPresentedPost(state, { postId, forGroupId: currentGroup?.id })
  const isProject = get('type', post) === 'project'

  return {
    postId,
    post,
    isProject,
    currentUser,
    currentGroup
  }
}

export function mapDispatchToProps (dispatch) {
  return { dispatch }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { postId, post } = stateProps
  const { dispatch } = dispatchProps
  const { navigation, route } = ownProps

  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    fetchPost: () => dispatch(fetchPost(postId)),
    joinProject: () => dispatch(joinProject(postId)),
    leaveProject: () => dispatch(leaveProject(postId)),
    editPost: () => navigation.navigate('Edit Post', { id: postId }),
    goToMembers: () => navigation.navigate('Project Members', { id: postId, members: get('members', post) }),
    showMember: id => {
      if (isModalScreen(route.name)) {
        return navigation.navigate(modalScreenName('Member'), { id })
      } else {
        return navigation.navigate('Member', { id })
      }
    },
    showTopic: topicName => navigation.navigate('Feed', { topicName }),
    respondToEvent: response => dispatch(respondToEvent(postId, response))
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps, { forwardRef: true })
