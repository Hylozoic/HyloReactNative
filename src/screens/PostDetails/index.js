import { withNavigationFocus } from '@react-navigation/compat'

import component from './PostDetails'
import connector from './PostDetails.connector'

export default withNavigationFocus(connector(component))
