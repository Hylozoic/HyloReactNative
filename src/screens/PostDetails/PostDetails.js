/* eslint-disable camelcase */
import React from 'react'
import { View, Text, TouchableOpacity, Alert } from 'react-native'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import { KeyboardAccessoryView } from '@flyerhq/react-native-keyboard-accessory-view'
import { get, isEmpty, find } from 'lodash/fp'
import { isIOS } from 'util/platform'
import useGroupSelect from 'hooks/useSelectGroup'
import SocketSubscriber from 'components/SocketSubscriber'
import Comments from 'components/Comments'
import PostBody from 'components/PostCard/PostBody'
import PostGroups from 'components/PostCard/PostGroups'
import PostImage from 'components/PostCard/PostImage'
import Files from 'components/Files'
import PostFooter from 'components/PostCard/PostFooter'
import PostHeader from 'components/PostCard/PostHeader'
import ProjectMembersSummary from 'components/ProjectMembersSummary'
import LoadingScreen from 'screens/LoadingScreen'
import Button from 'components/Button'
import InlineEditor, { toHtml } from 'components/InlineEditor'
import Icon from 'components/Icon'
import styles from './PostDetails.styles'

export class PostDetails extends React.Component {
  state = {
    replyingToName: null,
    commentText: '',
    reaplyingToCommentId: null
  }
  commentsRef = React.createRef()
  editorRef = React.createRef()

  componentDidMount () {
    this.props.fetchPost()
    this.setHeader()
  }
  
  componentDidUpdate(prevProps) {
    if (prevProps.currentGroup?.name !== this.props.currentGroup?.name) {
      this.setHeader()
    }
  }

  setHeader = () => {
    const { navigation, route, currentGroup } = this.props
    if (route?.name == 'Post Details - Modal') return
    navigation.setOptions({ title: currentGroup?.name  })
  }

  shouldComponentUpdate (nextProps) {
    return !!nextProps.isFocused
  }

  onShowTopic = (topicId) =>
    this.props.showTopic(topicId, get('post.groups.0.id', this.props))

  handleCreateComment = async commentText => {
    const commentTextAsHtml = toHtml(commentText)

    if (!isEmpty(commentTextAsHtml)) {
      this.setState(() => ({ submitting: true }))

      const { error } = await this.props.createComment({
        text: commentTextAsHtml,
        parentCommentId: this.state.reaplyingToCommentId
      })

      this.setState(() => ({ submitting: false }))

      if (error) {
        Alert.alert("Your comment couldn't be saved; please try again.")
      } else {
        this.handleCommentReplyCancel()
      }
    }
  }

  handleCommentOnChange = (commentText) => {
    this.setState(() => ({ commentText }))
  }

  handleCommentReplyCancel = callback => {
    this.setState({ replyingToName: null, commentText: '' }, () => {
      this.commentsRef?.current.highlightComment(null)
      this.editorRef?.editorInputRef.current.clear()
      this.editorRef?.editorInputRef.current.blur()
      callback && callback()
    })
  }

  handleCommentReply = (comment, { mention = false }) => {
    this.handleCommentReplyCancel(() => {
      this.setState({ replyingToName: comment.creator.name, commentText: '' })
      this.setState({ reaplyingToCommentId: comment.parentComment || comment.id })
  
      this.commentsRef?.current.highlightComment(comment)
      this.commentsRef?.current.scrollToComment(comment)
      this.editorRef?.editorInputRef.current.clear()
  
      if (mention) this.editorRef?.insertMention(comment.creator)
  
      this.editorRef?.editorInputRef.current.focus()  
    })
  }

  renderPostDetails = (panHandlers) => {
    const { post, currentUser, respondToEvent, showMember } = this.props
    const slug = get('groups.0.slug', post)
    const isMember = find(member => member.id === currentUser.id, post.members)
    const location = post.location || (post.locationObject && post.locationObject.fullText)

    return (
      <Comments style={styles.commentsScrollView}
        ref={this.commentsRef}
        postId={post.id}
        onReply={this.handleCommentReply}
        header={(
          <PostCardForDetails
            {...this.props}
            respondToEvent={respondToEvent}
            showTopic={this.onShowTopic}
            isMember={isMember}
            location={location}
          />
        )}
        slug={slug}
        showMember={showMember}
        showTopic={this.onShowTopic}
        panHandlers={panHandlers}
      />
    )
  }

  render () {
    const { post } = this.props
    const { commentText, replyingToName, submitting } = this.state
    const groupId = get('groups.0.id', post)

    if (!post?.creator || !post?.title) return <LoadingScreen />
    
    const isModal = this.props.route?.name == 'Post Details - Modal'

    return (
      <View style={styles.container}>
        <KeyboardAccessoryView
          contentContainerStyle={{
            ...styles.promptContentContainer,
            paddingBottom: isModal ? this.props.safeAreaInsets.bottom : 0
          }}
          // TODO: Calculate these?
          spaceBetweenKeyboardAndAccessoryView={isIOS ? -79 : 0}
          contentOffsetKeyboardOpened={isIOS ? -45 : 0}
          renderScrollable={this.renderPostDetails}>
            {replyingToName && (
              <View style={styles.commentPrompt}>
                <Text style={styles.commentPromptText}>
                  Replying to <Text style={{ fontWeight: 'bold' }}>
                    {replyingToName}</Text> {'\u00B7'} </Text>
                <TouchableOpacity onPress={() => this.handleCommentReplyCancel()}>
                  <Text style={styles.commentPromptClearLink}>Cancel</Text>
                </TouchableOpacity>
              </View>
            )}
            <InlineEditor
              style={styles.inlineEditor}
              onRef={elem => this.editorRef = elem}
              onChange={this.handleCommentOnChange}
              onSubmit={this.handleCreateComment}
              value={commentText}
              submitting={submitting}
              placeholder='Write a comment...'
              groupId={groupId}
            />
        </KeyboardAccessoryView>
        <SocketSubscriber type='post' id={post.id} />
      </View>
    )
  }
}

export default function (props) {
  useGroupSelect()
  const safeAreaInsets = useSafeAreaInsets()

  return <PostDetails {...props} safeAreaInsets={safeAreaInsets} />
}

export function PostCardForDetails ({
  currentUser,
  editPost,
  goToGroup,
  goToMembers,
  isProject,
  joinProject,
  leaveProject,
  post,
  respondToEvent,
  showMember,
  showTopic
}) {
  const slug = get('groups.0.slug', post)
  const isMember = find(member => member.id === currentUser.id, post.members)
  const location = post.location || (post.locationObject && post.locationObject.fullText)

  return (
    <View style={styles.postCard}>
      <PostHeader
        creator={post.creator}
        date={post.createdAt}
        type={post.type}
        editPost={editPost}
        groups={post.groups}
        slug={slug}
        pinned={post.pinned}
        topics={post.topics}
        showTopic={showTopic}
        postId={post.id}
        showMember={showMember}
        goToGroup={goToGroup}
        announcement={post.announcement}
        closeOnDelete
      />
      <PostImage imageUrls={post.imageUrls} linked />
      <PostBody
        details={post.details}
        endTime={post.endTime}
        linkPreview={post.linkPreview}
        myEventResponse={post.myEventResponse}
        respondToEvent={respondToEvent}
        showMember={showMember}
        showTopic={showTopic}
        slug={slug}
        startTime={post.startTime}
        title={post.title}
        type={post.type}
      />
      <Files urls={post.fileUrls} />
      {isProject && (
        <ProjectMembersSummary
          members={post.members}
          onPress={goToMembers}
          dimension={34}
          style={styles.projectMembersContainer}
        />
      )}
      {isProject && (
        <JoinProjectButton
          style={styles.projectJoinButton}
          leaving={isMember}
          onPress={isMember ? leaveProject : joinProject}
        />
      )}
      {!!location && (
        <View style={styles.infoRow}>
          <Icon style={styles.locationIcon} name='Location' />
          <Text style={styles.infoRowInfo} selectable>{location}</Text>
        </View>
      )}
      <PostGroups
        groups={post.groups}
        includePublic={post.isPublic}
        slug={slug}
        style={[styles.infoRow]}
        goToGroup={goToGroup}
        shouldShowGroups
      />
      <PostFooter
        style={styles.postFooter}
        id={post.id}
        type={post.type}
        currentUser={currentUser}
        commenters={post.commenters}
        commentersTotal={post.commentersTotal}
        members={post.members}
        eventInvitations={post.eventInvitations}
        votesTotal={post.votesTotal}
        myVote={post.myVote}
        showActivityLabel
      />
    </View>
  )
}

export function JoinProjectButton ({ style, onPress, leaving }) {
  const text = leaving ? 'Leave Project' : 'Join Project'
  return (
    <Button
      style={style}
      text={text}
      onPress={onPress}
    />
  )
}
