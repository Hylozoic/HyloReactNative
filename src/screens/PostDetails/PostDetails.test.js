import React from 'react'
import ReactShallowRenderer from 'react-test-renderer/shallow'
import TestRenderer, { act } from 'react-test-renderer'
import { Provider } from 'react-redux'
import { PostDetails, CommentPrompt, JoinProjectButton, ProjectMembers } from './PostDetails'
import { Linking, TouchableOpacity } from 'react-native'
import { createMockStore } from 'util/testing'
import MockedScreen from 'util/testing/MockedScreen'
import orm from 'store/models'

// TODO: Fix tests to have test Redux store for Redux hooks:
//       https://gist.github.com/krawaller/e5d40217658fa132f3c3904987e467cd

jest.mock('util/websockets', () => {
  const socket = {
    post: jest.fn(),
    on: jest.fn()
  }

  return {
    getSocket: () => Promise.resolve(socket),
    socketUrl: path => 'sockethost' + path
  }
})

const post = {
  id: '91',
  creator: {
    id: '77',
    name: 'Houdini'
  },
  groups: [{ slug: 'foom' }],
  createdAt: '2017-05-19T23:24:58Z',
  imageUrls: ['foom.png'],
  title: 'Hi',
  details: 'Lo',
  commenters: [{ id: 9, name: 'Jebobo Crustacean' }, { id: 7, name: 'Lobster Science' }],
  commentsTotal: 12,
  votesTotal: 8,
  myVote: true,
  type: 'request',
  fileUrls: [
    'http://foo.com/foo.pdf',
    'http://foo.com/bar.zip'
  ],
  pinned: true,
  topics: [
    { name: 'topic1', id: '1' }
  ]
}
const currentUser = {
  id: 123,
  avatarUrl: 'me.png',
  firstName: () => 'Joe'
}

const props = {
  post,
  currentUser,
  editPost: jest.fn(),
  pending: false,
  fetchPost: jest.fn(),
  showMember: jest.fn(),
  showTopic: jest.fn(),
  createComment: jest.fn(() => Promise.resolve({ success: true })),
  goToGroup: jest.fn(),
  navigation: {
    setOptions: jest.fn(),
    setParams: jest.fn(),
    getParam: jest.fn()
  }
}

const state = {
  orm: orm.getEmptyState(),
  queryResults: {},
  pending: {},
  session: {}
}

describe('PostDetails', () => {
  it('renders correctly', () => {
    const renderer = TestRenderer.create(
      <Provider store={createMockStore(state)}>
        <MockedScreen>
          {() => <PostDetails {...props} />}
        </MockedScreen>
      </Provider>
    )
    expect(renderer.toJSON()).toMatchSnapshot()
  })

  it('handleCreateComment success', async () => {
    const renderer = TestRenderer.create(
      <Provider store={createMockStore(state)}>
        <MockedScreen>
          {() => <PostDetails {...props} />}
        </MockedScreen>
      </Provider>
    )
    const instance = renderer.root.findByType(PostDetails).instance
    const commentText = 'some text [amention:0] #topic <some encoded stuff>'
    instance.setState({ commentText })
    await act(async () => (
      instance.handleCreateComment('some text [amention:3332] #topic <some encoded stuff>')
    ))
    expect(props.createComment).toHaveBeenCalledWith({ text: 'some text <a href="#" data-entity-type="mention" data-user-id="3332">amention</a> #topic &lt;some encoded stuff&gt;', "parentCommentId": null })
    expect(instance.state.submitting).toBeFalsy()
    expect(instance.state.commentText).toBe('')
  })

  it('handleCreateComment rejection', async () => {
    const rejectionProps = {
      ...props,
      createComment: jest.fn(() => Promise.resolve({ error: new Error('blah') }))
    }
    const renderer = TestRenderer.create(
      <Provider store={createMockStore(state)}>
        <MockedScreen>
          {() => <PostDetails {...rejectionProps} />}
        </MockedScreen>
      </Provider>
    )
    const instance = renderer.root.findByType(PostDetails).instance
    const commentText = 'some text [amention:0] #topic <some encoded stuff>'
    await act(async () => {
      await instance.setState({ commentText })
      await instance.handleCreateComment(commentText)
    })
    expect(instance.state.submitting).toBeFalsy()
    expect(instance.state.commentText).toBe(commentText)
  })

  it('handleCommentOnChange', async () => {
    const renderer = TestRenderer.create(
      <Provider store={createMockStore(state)}>
        <MockedScreen>
          {() => <PostDetails {...props} />}
        </MockedScreen>
      </Provider>
    )
    const instance = renderer.root.findByType(PostDetails).instance
    const commentText = 'some text [amention:0] #topic <some encoded stuff>'
    await act(async () => {
      await instance.setState({ commentText })
      await instance.handleCommentOnChange('something or nothing')
    })
    expect(instance.state.commentText).toEqual('something or nothing')
  })
})

describe('JoinProjectButton', () => {
  it('renders as expected', () => {
    const renderer = new ReactShallowRenderer()
    renderer.render(
      <JoinProjectButton onPress={() => {}} />
    )
    const actual = renderer.getRenderOutput()
    expect(actual).toMatchSnapshot()
  })

  it('renders as expected when leaving', () => {
    const renderer = new ReactShallowRenderer()
    renderer.render(
      <JoinProjectButton onPress={() => {}} leaving />
    )
    const actual = renderer.getRenderOutput()
    expect(actual).toMatchSnapshot()
  })
})
