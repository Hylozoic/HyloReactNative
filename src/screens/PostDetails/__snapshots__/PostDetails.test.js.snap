// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinProjectButton renders as expected 1`] = `
<Button
  onPress={[Function]}
  text="Join Project"
/>
`;

exports[`JoinProjectButton renders as expected when leaving 1`] = `
<Button
  onPress={[Function]}
  text="Leave Project"
/>
`;

exports[`PostDetails renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgb(242, 242, 242)",
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      collapsable={false}
      pointerEvents="box-none"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        onLayout={[Function]}
        pointerEvents="box-none"
        style={null}
      >
        <View
          collapsable={false}
          pointerEvents="box-none"
        >
          <View
            collapsable={false}
          />
        </View>
        <View
          collapsable={false}
          pointerEvents="box-none"
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 0,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "stretch",
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
            />
            <View
              collapsable={false}
              pointerEvents="box-none"
            >
              <Text
                accessibilityRole="header"
                aria-level="1"
                collapsable={false}
                numberOfLines={1}
                onLayout={[Function]}
              >
                MockedScreen
              </Text>
            </View>
            <View
              collapsable={false}
              pointerEvents="box-none"
            />
          </View>
        </View>
      </View>
    </View>
    <RNSScreenContainer
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        activityState={2}
        collapsable={false}
        forwardedRef={[Function]}
        pointerEvents="box-none"
      >
        <View
          collapsable={false}
        />
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onGestureEnd={[Function]}
          onOpen={[Function]}
          onTransition={[Function]}
          pointerEvents="box-none"
          style={
            Array [
              Object {
                "display": "flex",
                "overflow": undefined,
              },
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
            ]
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
          >
            <View
              collapsable={false}
              forwardedRef={[Function]}
              needsOffscreenAlphaCompositing={false}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
            >
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <View>
                      <View
                        collapsable={false}
                      >
                        <RCTScrollView
                          ListFooterComponent={[Function]}
                          contentContainerStyle={
                            Object {
                              "flexGrow": 1,
                              "justifyContent": "flex-end",
                            }
                          }
                          data={Array []}
                          getItem={[Function]}
                          getItemCount={[Function]}
                          initialScrollIndex={0}
                          invertStickyHeaders={true}
                          inverted={true}
                          keyExtractor={[Function]}
                          keyboardDismissMode="interactive"
                          onContentSizeChange={[Function]}
                          onLayout={[Function]}
                          onMomentumScrollBegin={[Function]}
                          onMomentumScrollEnd={[Function]}
                          onMoveShouldSetResponder={[Function]}
                          onMoveShouldSetResponderCapture={[Function]}
                          onResponderEnd={[Function]}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderReject={[Function]}
                          onResponderRelease={[Function]}
                          onResponderStart={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onScroll={[Function]}
                          onScrollBeginDrag={[Function]}
                          onScrollEndDrag={[Function]}
                          onStartShouldSetResponder={[Function]}
                          onStartShouldSetResponderCapture={[Function]}
                          renderItem={[Function]}
                          scrollEventThrottle={50}
                          stickyHeaderIndices={Array []}
                          style={
                            Array [
                              Object {
                                "transform": Array [
                                  Object {
                                    "scaleY": -1,
                                  },
                                ],
                              },
                              undefined,
                            ]
                          }
                        >
                          <View>
                            <View
                              onLayout={[Function]}
                            >
                              <View>
                                <View>
                                  <View>
                                    <View
                                      accessible={true}
                                      collapsable={false}
                                      focusable={true}
                                      onClick={[Function]}
                                      onResponderGrant={[Function]}
                                      onResponderMove={[Function]}
                                      onResponderRelease={[Function]}
                                      onResponderTerminate={[Function]}
                                      onResponderTerminationRequest={[Function]}
                                      onStartShouldSetResponder={[Function]}
                                    />
                                  </View>
                                  <View>
                                    <View
                                      accessible={true}
                                      collapsable={false}
                                      focusable={true}
                                      onClick={[Function]}
                                      onResponderGrant={[Function]}
                                      onResponderMove={[Function]}
                                      onResponderRelease={[Function]}
                                      onResponderTerminate={[Function]}
                                      onResponderTerminationRequest={[Function]}
                                      onStartShouldSetResponder={[Function]}
                                    >
                                      <Text>
                                        Houdini
                                      </Text>
                                    </View>
                                    <View>
                                      <Text>
                                        4 years ago
                                      </Text>
                                      <RCTScrollView
                                        data={
                                          Array [
                                            Object {
                                              "id": "1",
                                              "name": "topic1",
                                            },
                                          ]
                                        }
                                        getItem={[Function]}
                                        getItemCount={[Function]}
                                        horizontal={true}
                                        keyExtractor={[Function]}
                                        onContentSizeChange={[Function]}
                                        onLayout={[Function]}
                                        onMomentumScrollBegin={[Function]}
                                        onMomentumScrollEnd={[Function]}
                                        onScroll={[Function]}
                                        onScrollBeginDrag={[Function]}
                                        onScrollEndDrag={[Function]}
                                        removeClippedSubviews={false}
                                        renderItem={[Function]}
                                        scrollEventThrottle={50}
                                        showsHorizontalScrollIndicator={false}
                                        stickyHeaderIndices={Array []}
                                        viewabilityConfigCallbackPairs={Array []}
                                      >
                                        <View>
                                          <View
                                            onLayout={[Function]}
                                          >
                                            <View
                                              accessible={true}
                                              collapsable={false}
                                              focusable={true}
                                              onClick={[Function]}
                                              onResponderGrant={[Function]}
                                              onResponderMove={[Function]}
                                              onResponderRelease={[Function]}
                                              onResponderTerminate={[Function]}
                                              onResponderTerminationRequest={[Function]}
                                              onStartShouldSetResponder={[Function]}
                                            >
                                              <Text>
                                                #
                                                topic1
                                              </Text>
                                            </View>
                                          </View>
                                        </View>
                                      </RCTScrollView>
                                    </View>
                                  </View>
                                  <View>
                                    <Icon
                                      name="Pin"
                                    />
                                    <View>
                                      <Text>
                                        REQUEST
                                      </Text>
                                    </View>
                                    <View
                                      accessible={true}
                                      collapsable={false}
                                      focusable={true}
                                      hitSlop={
                                        Object {
                                          "bottom": 10,
                                          "left": 10,
                                          "right": 15,
                                          "top": 20,
                                        }
                                      }
                                      onClick={[Function]}
                                      onResponderGrant={[Function]}
                                      onResponderMove={[Function]}
                                      onResponderRelease={[Function]}
                                      onResponderTerminate={[Function]}
                                      onResponderTerminationRequest={[Function]}
                                      onStartShouldSetResponder={[Function]}
                                    >
                                      <Icon
                                        name="More"
                                      />
                                    </View>
                                  </View>
                                </View>
                                <View>
                                  <View
                                    accessible={true}
                                    collapsable={false}
                                    focusable={true}
                                    onClick={[Function]}
                                    onResponderGrant={[Function]}
                                    onResponderMove={[Function]}
                                    onResponderRelease={[Function]}
                                    onResponderTerminate={[Function]}
                                    onResponderTerminationRequest={[Function]}
                                    onStartShouldSetResponder={[Function]}
                                  >
                                    <View
                                      accessibilityIgnoresInvertColors={true}
                                    >
                                      <Image
                                        source={
                                          Object {
                                            "uri": "foom.png",
                                          }
                                        }
                                      />
                                    </View>
                                  </View>
                                </View>
                                <View>
                                  <View>
                                    <Text>
                                      Hi
                                    </Text>
                                  </View>
                                  <View>
                                    <Text
                                      style={
                                        Object {
                                          "color": "#5D757A",
                                          "fontFamily": "Circular-Book",
                                          "fontSize": 14,
                                          "lineHeight": 20,
                                          "marginBottom": 8,
                                        }
                                      }
                                    >
                                      Lo
                                    </Text>
                                  </View>
                                </View>
                                <View>
                                  <View
                                    accessible={true}
                                    collapsable={false}
                                    focusable={true}
                                    onClick={[Function]}
                                    onResponderGrant={[Function]}
                                    onResponderMove={[Function]}
                                    onResponderRelease={[Function]}
                                    onResponderTerminate={[Function]}
                                    onResponderTerminationRequest={[Function]}
                                    onStartShouldSetResponder={[Function]}
                                  >
                                    <View>
                                      <Icon
                                        name="Document"
                                      />
                                      <Text
                                        numberOfLines={2}
                                      >
                                        foo.pdf
                                      </Text>
                                    </View>
                                  </View>
                                  <View
                                    accessible={true}
                                    collapsable={false}
                                    focusable={true}
                                    onClick={[Function]}
                                    onResponderGrant={[Function]}
                                    onResponderMove={[Function]}
                                    onResponderRelease={[Function]}
                                    onResponderTerminate={[Function]}
                                    onResponderTerminationRequest={[Function]}
                                    onStartShouldSetResponder={[Function]}
                                  >
                                    <View>
                                      <Icon
                                        name="Document"
                                      />
                                      <Text
                                        numberOfLines={2}
                                      >
                                        bar.zip
                                      </Text>
                                    </View>
                                  </View>
                                </View>
                                <View>
                                  <View>
                                    <Text>
                                      Posted In: 
                                    </Text>
                                    <View>
                                      <View
                                        accessible={true}
                                        collapsable={false}
                                        focusable={true}
                                        onClick={[Function]}
                                        onResponderGrant={[Function]}
                                        onResponderMove={[Function]}
                                        onResponderRelease={[Function]}
                                        onResponderTerminate={[Function]}
                                        onResponderTerminationRequest={[Function]}
                                        onStartShouldSetResponder={[Function]}
                                      >
                                        <Text
                                          numberOfLines={1}
                                        />
                                      </View>
                                    </View>
                                    <View
                                      accessible={true}
                                      collapsable={false}
                                      focusable={true}
                                      onClick={[Function]}
                                      onResponderGrant={[Function]}
                                      onResponderMove={[Function]}
                                      onResponderRelease={[Function]}
                                      onResponderTerminate={[Function]}
                                      onResponderTerminationRequest={[Function]}
                                      onStartShouldSetResponder={[Function]}
                                    >
                                      <Icon
                                        name="ArrowDown"
                                      />
                                    </View>
                                  </View>
                                </View>
                                <Text>
                                  ACTIVITY
                                </Text>
                                <View>
                                  <View>
                                    <View>
                                      <Image
                                        source={
                                          Object {
                                            "testUri": "../../../assets/default-user-avatar.png",
                                          }
                                        }
                                      />
                                    </View>
                                    <View>
                                      <Image
                                        source={
                                          Object {
                                            "testUri": "../../../assets/default-user-avatar.png",
                                          }
                                        }
                                      />
                                    </View>
                                    <Text>
                                      Jebobo and Lobster commented
                                    </Text>
                                  </View>
                                  <View
                                    accessible={true}
                                    collapsable={false}
                                    focusable={true}
                                    onClick={[Function]}
                                    onResponderGrant={[Function]}
                                    onResponderMove={[Function]}
                                    onResponderRelease={[Function]}
                                    onResponderTerminate={[Function]}
                                    onResponderTerminationRequest={[Function]}
                                    onStartShouldSetResponder={[Function]}
                                  >
                                    <Icon
                                      name="ArrowUp"
                                    />
                                    <Text>
                                      8
                                    </Text>
                                  </View>
                                </View>
                              </View>
                            </View>
                          </View>
                        </RCTScrollView>
                      </View>
                      <View
                        collapsable={false}
                        testID="container"
                      >
                        <View
                          onLayout={[Function]}
                          style={
                            Object {
                              "backgroundColor": "#f4f5f7",
                              "borderWidth": 0,
                              "flex": 1,
                              "marginBottom": 0,
                              "marginLeft": 0,
                              "marginRight": 0,
                              "paddingBottom": 0,
                            }
                          }
                        >
                          <View
                            onBlur={[Function]}
                            onFocus={[Function]}
                          >
                            <View>
                              <TextInput
                                editable={true}
                                multiline={true}
                                onChangeText={[Function]}
                                onSelectionChange={[Function]}
                                placeholder="Write a comment..."
                                placeholderTextColor="#c0c6cd"
                                underlineColorAndroid="transparent"
                                value=""
                              />
                              <View>
                                <View
                                  accessible={true}
                                  collapsable={false}
                                  focusable={true}
                                  hitSlop={
                                    Object {
                                      "bottom": 7,
                                      "left": 7,
                                      "right": 7,
                                      "top": 7,
                                    }
                                  }
                                  onClick={[Function]}
                                  onResponderGrant={[Function]}
                                  onResponderMove={[Function]}
                                  onResponderRelease={[Function]}
                                  onResponderTerminate={[Function]}
                                  onResponderTerminationRequest={[Function]}
                                  onStartShouldSetResponder={[Function]}
                                >
                                  <Text>
                                    @
                                  </Text>
                                </View>
                                <View
                                  accessible={true}
                                  collapsable={false}
                                  focusable={true}
                                  hitSlop={
                                    Object {
                                      "bottom": 7,
                                      "left": 7,
                                      "right": 7,
                                      "top": 7,
                                    }
                                  }
                                  onClick={[Function]}
                                  onResponderGrant={[Function]}
                                  onResponderMove={[Function]}
                                  onResponderRelease={[Function]}
                                  onResponderTerminate={[Function]}
                                  onResponderTerminationRequest={[Function]}
                                  onStartShouldSetResponder={[Function]}
                                >
                                  <Text>
                                    #
                                  </Text>
                                </View>
                              </View>
                            </View>
                            <View
                              accessibilityState={
                                Object {
                                  "disabled": true,
                                }
                              }
                              accessible={true}
                              collapsable={false}
                              focusable={true}
                              hitSlop={
                                Object {
                                  "bottom": 10,
                                  "left": 10,
                                  "right": 10,
                                  "top": 5,
                                }
                              }
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                            >
                              <Text
                                allowFontScaling={false}
                                style={
                                  Array [
                                    Object {
                                      "color": undefined,
                                      "fontSize": 26,
                                    },
                                    undefined,
                                    Object {
                                      "fontFamily": "Material Icons",
                                      "fontStyle": "normal",
                                      "fontWeight": "normal",
                                    },
                                    Object {},
                                  ]
                                }
                              >
                                
                              </Text>
                            </View>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenContainer>
  </View>
</View>
`;
